src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php

```diff
@@ -0,0 +1,227 @@
+     */
+    public function indexAction(CurrencyFilters $filters)
+    {
+        $currencyGridFactory = $this->get('prestashop.core.grid.factory.currency');
+        $currencyGrid = $currencyGridFactory->getGrid($filters);
+        $gridPresenter = $this->get('prestashop.core.grid.presenter.grid_presenter');
+
+        return [
+            'currencyGrid' => $gridPresenter->present($currencyGrid),
+        ];
+    }
+
+    /**
+     * Provides filters functionality.
+     *
+     * @param Request $request
+     *
+     * @return RedirectResponse
+     */
+    public function searchAction(Request $request)
+    {
+        $definitionFactory = $this->get('prestashop.core.grid.definition.factory.currency');
+        $definitionFactory = $definitionFactory->getDefinition();
+
+        $gridFilterFormFactory = $this->get('prestashop.core.grid.filter.form_factory');
+        $searchParametersForm = $gridFilterFormFactory->create($definitionFactory);
+        $searchParametersForm->handleRequest($request);
+
+        $filters = [];
+        if ($searchParametersForm->isSubmitted()) {
+            $filters = $searchParametersForm->getData();
+        }
+
+        return $this->redirectToRoute('admin_currencies_index', ['filters' => $filters]);
+    }
+
+    /**
+     * Displays currency form.
+     *
+     * @AdminSecurity("is_granted('create', request.get('_legacy_controller'))", message="You do not have permission to add this.")
+     *
+     * @return RedirectResponse
+     */
+    public function createAction()
+    {
+        //todo: drop legacy and after having post add @DemoRestricted
+        $legacyLink = $this->getAdminLink('AdminCurrencies', [
+            'addcurrency' => 1,
+        ]);
+
+        return $this->redirect($legacyLink);
+    }
+
+    /**
+     * Displays currency form.
+     *
+     * @AdminSecurity("is_granted('update', request.get('_legacy_controller'))", message="You do not have permission to edit this.")
+     *
+     * @param int $currencyId
+     *
+     * @return RedirectResponse
+     */
+    public function editAction($currencyId)
+    {
+        //todo: drop legacy and after having post add @DemoRestricted
+        $legacyLink = $this->getAdminLink('AdminCurrencies', [
+            'id_currency' => $currencyId,
+            'updatecurrency' => 1,
+        ]);
+
+        return $this->redirect($legacyLink);
+    }
+
+    /**
+     * Deletes currency.
+     *
+     * @AdminSecurity("is_granted('delete', request.get('_legacy_controller'))", message="You do not have permission to delete this.")
+     * @DemoRestricted(redirectRoute="admin_currencies_index")
+     *
+     * @param int $currencyId
+     *
+     * @return RedirectResponse
+     */
+    public function deleteAction($currencyId)
+    {
+        try {
+            $currencyId = new CurrencyId($currencyId);
+            $this->getCommandBus()->handle(new DeleteCurrencyCommand($currencyId));
+        } catch (CurrencyException $exception) {
```

Comment:
What happens if another Exception is raised, but not a `CurrencyException` ? Fatal error ? Then what happens in the browser ? If this causes a JS error that crashes the JS in browser, we need to handle that.

Reply:
Currently , the `CurrencyException` handles also `PrestaShopException` which has a lot of scenarios covered when working with legacy ObjectModels inside the command. It also covers `CurrencyNotFoundException`, `CannotDeleteCurrencyException`, `CannotDeleteDefaultCurrencyException` etc...

I'm not sure I understand the problem, but do you suggest to handle exceptions like this:
```php
        try {
            $currencyId = new CurrencyId($currencyId);
            $this->getCommandBus()->handle(new DeleteCurrencyCommand($currencyId));
        } catch (Exception $exception) {}
```
the main difference is that now I can handle every exception and display beautiful non informative error to regular user if an unexpected error occurred.  But it might be harder for others to search and identify the problem :thinking:

Reply:
@tomas862 You got me convinced 👍